### PR TITLE
:tada: Snowflake Destination  internal staging support

### LINKED ISSUE #4370: Snowflake destination: Support internal staging
## Tell us about the problem you're trying to solve
Records can be inserted into snowflake using a standard SQL `INSERT` statement or by first copying the data to blob storage (e.g: azure, gcs, s3, snowflake internal) then using the `COPY` command to read that data into snowflake. The staging approach is an order of magnitude faster and should pretty much always be used except for the smallest data volumes/proof of concepts. 

The snowflake destination currently supports staging to S3. it should also support internal staging as described here: https://docs.snowflake.com/en/user-guide/data-load-considerations-stage.html

## Describe the solution you’d like
Support internal staging in snowflake. 

## Acceptance criteria
1. Internal staging is supported
2. All docs are updated with explicit documentation of needed DB user permissions
3.  If no backwards incompatible changes are introduced in terms of permissions, remove the INSERT mode in a backwards compatible way. If we have internal staging then there is basically no reason to use inserts

### PR DESCRIPTION
**Pull request by VitaliiMaltsev**
## What
Records can be inserted into snowflake using a standard SQL INSERT statement or by first copying the data to blob storage (e.g: azure, gcs, s3, snowflake internal) then using the COPY command to read that data into snowflake. The staging approach is an order of magnitude faster and should pretty much always be used except for the smallest data volumes/proof of concepts.

The snowflake destination currently supports staging to S3. it should also support internal staging as described here: https://docs.snowflake.com/en/user-guide/data-load-considerations-stage.html

## How
Workload :
CREATE STAGE @TEMP_STAGE_NAME
PUT file://local/some_folder/* @TEMP_STAGE_NAME. 
COPY FROM @TEMP_STAGE_NAME
DROP @TEMP_STAGE_NAME

Standard Inserts 3 millions records
snowflake_insert_3m
Internal Staging  3 millions records
snowflake_int_3m_7

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? If yes, please make sure to include it here and in any changelogs with the 🚨🚨 emoji
What is the end result perceived by the user?

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated 
    - [ ] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by andriikorotkov — APPROVED** [review]

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingDestinatiomAcceptanceTest.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+
+public class SnowflakeInternalStagingDestinatiomAcceptanceTest extends DestinationAcceptanceTest {
```
Comment (tuliren) [inline airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingDestinatiomAcceptanceTest.java:36]: This file is very similar to `SnowflakeInsertDestinationAcceptanceTest`. Can this class just extend from `SnowflakeInsertDestinationAcceptanceTest` and override the few different methods? Both the GCS and S3 acceptance tests already do that.

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingConsumerFactory.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SnowflakeInternalStagingConsumerFactory {
```
Comment (tuliren) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingConsumerFactory.java:34]: I just realized that this class is kind of similar to the `JdbcBufferedConsumerFactory`. Is there any way to reuse the code? Totally fine if it is not. I just want to make sure that we have considered this route.

Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingConsumerFactory.java:34]: > 

That was my first suggestion to use JdbcBufferedConsumerFactory, but internal staging uses some phases specific only for Snowflake, that's why i created a new class

Comment (tuliren) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingConsumerFactory.java:43]: Got it. Please ignore my first comment then.

File: airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json [hunk 3]
```diff
@@ -200,6 +200,19 @@
...
+            "additionalProperties": false,
+            "description": "Writes large batches of records to a file, uploads the file to Snowflake, then uses <pre>COPY INTO table</pre> to upload the file. Recommended for large production workloads for better speed and scalability.",
+            "required": ["method"],
+            "properties": {
+              "method": {
+                "type": "string",
+                "enum": ["Internal Staging"],
+                "default": "Internal Staging"
+              }
+            }
```
[the remark this answers is not in the document]
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json]: moved to the first element and also removed Standard Inserts from the list

File: airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json [hunk 4]
```diff
@@ -200,6 +200,19 @@
                 "order": 3
               }
             }
+          },
+          {
+            "title": "Internal Staging",
```
[the remark this answers is not in the document]
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json]: done

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingConsumerFactory.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SnowflakeInternalStagingConsumerFactory.class);
+
+  private static final int MAX_BATCH_SIZE_BYTES = 1024 * 1024 * 1024 / 4; // 256mb
```
[the remark this answers is not in the document]
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingConsumerFactory.java]: done

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeStagingSqlOperations.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    if (records.isEmpty()) {
+      return;
+    }
+      try {
+        loadDataIntoStage(database, stage, records);
+      } catch (Exception e) {
+        LOGGER.error("Failed to upload records into stage {}", stage, e);
+        throw new RuntimeException(e);
+      }
```
[the remark this answers is not in the document]
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeStagingSqlOperations.java]: formatted

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingDestinatiomAcceptanceTest.java (hunk 1 above)
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingDestinatiomAcceptanceTest.java]: done

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingConsumerFactory.java (hunk 2 above)
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeInternalStagingConsumerFactory.java:43]: added javadoc

**Review by tuliren — APPROVED** [review]
The test failure does not look relevant to this PR. The issue should have been fixed already on the master branch. Maybe you just need to merge master and the build should pass.

**Comment by VitaliiMaltsev** [comment]
Can not publish new version at the moment. This PR https://github.com/airbytehq/airbyte/pull/5814/files broke normalization step

**Pull request merged**